api/controller/spotify_controller.js

```diff
@@ -69,22 +69,143 @@ async function bulkFetchRandomizedItems(endpoint, access_token, objs, batch_limi
+  let selectedRandoms = []
+  let weight_sum = 0;
+  let total_size = Object.keys(data_struct.items).length; // original size of data_struct
+
+  // Sum up total weights of all items
+  for (let key in data_struct.items) {
+    weight_sum += data_struct.items[key].weight
+  }
+
+  // If the data_struct has more or equal items than max_size, loop while the number of selected is less than max_size
+  // If the data_struct has less items than max_size, loop while the number of selected is less than the number of items in data_struct
+  while ((total_size >= max_size && selectedRandoms.length < max_size) || (total_size < max_size && selectedRandoms.length < total_size)) {
+    // Pick a weight in range [0, weight_sum)
+    var rand_weight = Math.floor(Math.random() * (weight_sum));
+    var curr_weight = 0;
+    var chosen_id;
+
+    // For each item, sum their weights
+    for (var key in data_struct.items) {
+      curr_weight += data_struct.items[key].weight;
+      if (curr_weight > rand_weight) {
+        // Select the current item when the total item weights so far is greater than random weight
+        selectedRandoms.push(key);
+        chosen_id = key;
+        weight_sum -= data_struct.items[key].weight;
+        break;
+      }
+    }
+
+    // Remove the added item from the pool
+    delete data_struct.items[chosen_id];
+  }
+
+  return selectedRandoms;
+}
+
 module.exports = {
+  userSeedRecommendedSongSelection: async (req, res) => {
+    const { access_token, limit: max_result_tracks } = req.body;
+    const max_seed_list_size = 5;
+
+    const spotify_user_data = await USER_MANAGER.fetchUserData(req.body.access_token);
+    if (spotify_user_data.error != null) return res.json(spotify_user_data);
+
+    const seeds = await USER_MANAGER.fetchUserSeeds(spotify_user_data.email);
+    if (seeds.error != null) return res.json(seeds);
+
+    let artist_seeds = selectWeightedRandoms(seeds.fav_artists, max_seed_list_size);
+    let genre_seeds = selectWeightedRandoms(seeds.fav_genres, max_seed_list_size);
+
+    // Combined the seeds, shuffle and pick 5
+    let combined_seeds = getRandomSublist(artist_seeds.concat(genre_seeds), 5);
+
+    // Now that we have the shuffled seeds, redistribute them back to their types so we know which ones are ids and which ones are genres
+    artist_seeds = [];
+    genre_seeds = [];
+    for (let i = 0; i < combined_seeds.length; i++) {
+      if (SPOTIFY_UTILS.isValidGenreSeed(combined_seeds[i])) {
+        genre_seeds.push(combined_seeds[i]);
+      } else {
+        artist_seeds.push(combined_seeds[i]);
+      }
+    }
+
+    // Make the seed recommendation api request
+    let user_country = spotify_user_data.country;
+    artist_seeds = artist_seeds.join(",");
+    genre_seeds = genre_seeds.join(",");
+
+    const options = {
+      url: `https://api.spotify.com/v1/recommendations?limit=30&market=${user_country}&seed_artists=${artist_seeds}&seed_genres=${genre_seeds}&min_popularity=30`,
```

Comment:
Take out of options and append if seed_artists and genre_seeds are not empty strings, etc